### PR TITLE
Implementation of 3D data_override as well as data nudge procedure, w…

### LINKED ISSUE #125: FMS coupler request: blending capability for data_override
To support a flux-adjusted version of SPEAR, as well as many other future experiments throughout the lab, we'd like to request addition of a "data blending" capability to the FMS coupler.  This could be done as a simple extension of the existing data_override code, and would enable much greater control over the fluxes and fields passed between model components.  The proposal is as follows.

For each model field to be blended -- call it "var", e.g. sea surface temperature (SST) or zonal surface wind stress (tau_x) -- the user would provide up to three NetCDF files & variables:

a) var_data.nc, containing  the {lon,lat,time}-dimensioned variable "var": input data, e.g. satellite-observed surface zonal wind stresses (tau_x) in MKS units.  If var_data.nc were not provided, then the data would default to 0 everywhere.

b) var_data_coeff.nc, containing the {lon,lat}-dimensioned variable "coeff": a static field of coefficients to multiply the data read in from var_data.nc.  If this NetCDF file were not provided, then the data_coeff would default to 0 everywhere.

c)  var_coupler_coeff.nc, containing the {lon,lat}-dimensioned variable "coeff": a static field of coefficients to multiply the model-generated field at every coupling time step -- e.g. the tau_x produced by the atmospheric model.  If this NetCDF file were not provided, then the coupler_coeff would default to (1 - data_coeff) everywhere, easily allowing multi-region and tapering capability.

All three input variables (data, data_coeff, coupler_coeff) would consist of floating point values in the range (-inf,inf).

It would be the user's responsibility to provide the appropriate input data and coefficients, with any tapering etc. already applied.  So all the regional/blending magic would be pre-computed by the user.  The FMS coupler would just read the files, and then compute a simple weighted sum at each timestep.

Specifically: At initialization, the coupler would read and interpolate the data_coeff and coupler_coeff to the model's {lon,lat} grid.  Then during the model run, at each time step the coupler would read and interpolate the input data to the model {lon,lat,time} grid, using the existing time_interp_external() code which knows how to deal with calendars and modulo (climatological) time axes.  The coupler would then modify the var passed between model components (e.g. tau_x passed from atmos to ocean, or SST passed from ocean to atmos) as:

   var_blend = (var_coupler * var_coupler_coeff) + (var_data * var_data_coeff)

The coupler would also register the following diagnostics with the diag_manager, so that they could be requested via the diag_table:

   var_coupler_part = var_coupler * var_coupler_coeff
   var_data_part = var_data * var_data_coeff

The user could specify blends for more than one variable at a time -- e.g. SST, tau_x, tau_y, heat flux, and freshwater flux -- each with their own coupler_coeff and data_coeff.  This could be implemented using a data_blend_table (akin to a field_table).

This simple blending feature would enable and unify some powerful new capabilities that are not available with the existing data_override.  These include:

1) A spatially-varying override, with regional tapering from data toward model.  To do this the user would simply provide the input data, and their pre-computed spatial map of data coefficients:
   data_coeff = [0.0 .. 1.0]
It would be the user's responsibility to compute and provide the desired {lon,lat} map of data coefficients.  In this case the model_coeff.nc file would not be needed, and would default to 1.0-data_coeff.  At points where data_coeff=0.0, the model field would pass through unaltered.  At points where data_coeff=1.0, the model field would be completely overridden by the input data.  And at points where data_coeff=0.5, the result would be a mixture of half model and half data.

The existing data_override code only allows a 100% override in a rectangular lon-lat region specified via a namelist.  This old functionality could be replicated by the user setting data_coeff=1.0 in a rectangular {lon,lat} override region, setting data_coeff=0.0 everywhere else.  If need be, the existing namelist region-specification options could be retained for backward compatibility, or to provide a simpler interface in these special cases.

2) A spatially-varying "add" (this is a new feature).  For example, this could be used to impose a wind stress perturbation *in addition to* the model-generated stress, either to apply a climatological wind stress correction (flux adjustment), or to apply a transient regional wind burst to excite an El Nino event.  The user would provide the appropriate weightings:
   data_coeff = [0.0 .. 1.0]
   coupler_coeff = 1.0 everywhere

3) Override and add simultaneously (another new feature).  For example, this could be used to totally override the wind stress over the Indian & Atlantic oceans, and also add a climatological wind stress correction in the tropical Pacific that tapers toward zero in mid-latitudes.  These regions could be the same, different, or even partially overlapping; the user would just need to provide the appropriate coefficients for model and data.  The user would provide:
   data_coeff = [0.0 .. 1.0]
   coupler_coeff = [1.0 .. 1.0-data_coeff]

If at all possible, we would like to begin using these capabilities in October 2019.  Please let us know if any further information is needed to move forward with implementation.

Thank you for your help,

Andrew, Fanrong, Will, and Matt

PS: We had considered combining var_data*var_data_coeff into a single XYT input variable, but decided against it.  The var_data.nc will typically be a big XYT obs data file, while the var_data_coeff will be a small XY file.  If the coefficients ever needed to be modified, it'd be much easier if it was its own little entity.

### PR DESCRIPTION
**Pull request by GFDL-Eric**
**Description**
Implementation of 3D data_override (a one line fix in time_interp_external2), as well as the data nudge procedure (was data_blend), with accompanying changes in the test system.

Fixes #125 

**How Has This Been Tested?**
Tested on Gaea with Uriel's lm4p1_reg.xml (mainly using 2019.01 tag for various repos), rts regression test with prod,openmp (successfully completes) and repro,openmp (successfully reproduces answers), fre:bronx-16, Intel16, c4.

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [ ] `make distcheck` passes

### CODE REVIEW

File: data_override/data_override.F90 [hunk 1]
```diff
@@ -605,6 +633,84 @@ subroutine get_domainUG(gridname, UGdomain, comp_domain)
      call mpp_get_compute_domain(SGdomain,comp_domain(1),comp_domain(2),comp_domain(3),comp_domain(4))
 end subroutine get_domainUG
 !===============================================================================================
+! <FUNCTION NAME="new_nudge3D">
+!   <DESCRIPTION>
+! This function constructs a 3D data_nudge object.
+!   </DESCRIPTION>
```
Comment (thomas-robinson) [inline data_override/data_override.F90]: Comments should be in doxygen style

File: data_override/data_override.F90 [hunk 2]
```diff
@@ -605,6 +633,84 @@ subroutine get_domainUG(gridname, UGdomain, comp_domain)
...
+function new_nudge3D(newgridname, newvarname, xbounds, ybounds, zbounds, newobscoeffname, newmodcoeffname)
```
Comment (thomas-robinson) [inline data_override/data_override.F90]: The function name `new_nudge3d` is sort of relative.  This won't be so "new" in 10 years.  `nudge3d_construct` might be a better choice of naming as it is more descriptive and is not relative to the current time.

File: data_override/data_override.F90 [hunk 3]
```diff
@@ -115,6 +115,29 @@ module data_override_mod
...
+type nudge3D
+   character(len=3)                 :: gridname
+   character(len=128)               :: varname
+   character(len=128)               :: obs_coefficient_name
+   character(len=128)               :: mod_coefficient_name
+   real, _ALLOCATABLE               :: model_input(:,:,:) _NULL
+   real, _ALLOCATABLE               :: nudge_output(:,:,:) _NULL
+   real, _ALLOCATABLE               :: observed_data(:,:,:) _NULL
+   real, _ALLOCATABLE               :: observation_coefficient(:,:,:) _NULL
+   real, _ALLOCATABLE               :: model_coefficient(:,:,:) _NULL
```
Comment (thomas-robinson) [inline data_override/data_override.F90]: IMHO, these `_ALLOCATABLE` and `_NULL` should just be allocatable and no null is needed
```fortran
    real, ALLOCATABLE               :: model_input(:,:,:) 
￼   real, ALLOCATABLE               :: nudge_output(:,:,:) 
￼   real, ALLOCATABLE               :: observed_data(:,:,:) 
￼   real, ALLOCATABLE               :: observation_coefficient(:,:,:) 
￼   real, ALLOCATABLE               :: model_coefficient(:,:,:) 
```

File: data_override/data_override.F90 [hunk 4]
```diff
@@ -605,6 +633,84 @@ subroutine get_domainUG(gridname, UGdomain, comp_domain)
...
+   allocate(new_nudge3D%observed_data(xbounds(1):xbounds(2),ybounds(1):ybounds(2),zbounds(1):zbounds(2)))
+   allocate(new_nudge3D%observation_coefficient(xbounds(1):xbounds(2),ybounds(1):ybounds(2),zbounds(1):zbounds(2)))
+   allocate(new_nudge3D%model_coefficient(xbounds(1):xbounds(2),ybounds(1):ybounds(2),zbounds(1):zbounds(2)))
+
+end function new_nudge3D
+!===============================================================================================
+! <SUBROUTINE NAME="data_nudge_3d">
+!   <DESCRIPTION>
+! This routine performs data nudge for a variable, its observations, and the associated coefficients.
+!   </DESCRIPTION>
```
Comment (thomas-robinson) [inline data_override/data_override.F90]: These comments should be in doxygen format

File: data_override/data_override.F90 [hunk 5]
```diff
@@ -115,6 +115,29 @@ module data_override_mod
...
+type nudge2D
+   character(len=3)                 :: gridname
+   character(len=128)               :: varname
+   character(len=128)               :: obs_coefficient_name
+   character(len=128)               :: mod_coefficient_name
+   real, _ALLOCATABLE               :: model_input(:,:) _NULL
+   real, _ALLOCATABLE               :: nudge_output(:,:) _NULL
+   real, _ALLOCATABLE               :: observed_data(:,:) _NULL
+   real, _ALLOCATABLE               :: observation_coefficient(:,:) _NULL
+   real, _ALLOCATABLE               :: model_coefficient(:,:) _NULL
```
Comment (thomas-robinson) [inline data_override/data_override.F90]: This is the same comment I have above about the `_ALLOCATABLE` and `_NULL` statements.  They should be consistent at least.

File: test_fms/data_override/data_table_base [hunk 6]
```diff
@@ -2,3 +2,6 @@
 "ICE", "sic_obs",  "SIC", "INPUT/sst_ice_clim.nc", .false., 300.0
 "OCN", "sst_obs",  "SST", "INPUT/sst_ice_clim.nc", .false., 300.0
 "LND", "sst_obs",  "SST", "INPUT/sst_ice_clim.nc", .false., 300.0
+"ATM", "so2_cont_volc", "SO2_cont_volc", "INPUT/so2_with_coeffs.nc", .false., 1.0
+"ATM", "so2_cont_volc_coeff", "SO2_cont_volc_coeff", "INPUT/so2_with_coeffs.nc", .false., 1.0
+"ATM", "so2_cont_volc_mod_coeff", "", "", .false., 0.8
```
Comment (thomas-robinson) [inline test_fms/data_override/data_table_base]: Are these updates required for the 3D implementation and data nudge procedure?  These additions seem like they should be mentioned separately in a commit message.

File: test_fms/data_override/test_data_override2.sh [hunk 7]
```diff
@@ -22,3 +22,10 @@ tnum=$( printf "%2.2d" 2 )
 sed "s/<test_num>/${tnum}/"  $top_srcdir/test_fms/data_override/input_base.nml > input.nml
 run_test test_data_override 2 skip
 
+tnum=$( printf "%2.2d" 3 )
+sed "s/<test_num>/${tnum}/"  $top_srcdir/test_fms/data_override/input_base.nml > input.nml
+srun -n 1 ./test_data_override
+
+tnum=$( printf "%2.2d" 4 )
+sed "s/<test_num>/${tnum}/"  $top_srcdir/test_fms/data_override/input_base.nml > input.nml
+srun -n 1 ./test_data_override_3D
```
Comment (thomas-robinson) [inline test_fms/data_override/test_data_override2.sh]: This should be `run_test` instead of `srun`

File: data_override/data_override.F90 [hunk 8]
```diff
@@ -605,6 +633,84 @@ subroutine get_domainUG(gridname, UGdomain, comp_domain)
...
+   call data_override_3d(nudgeobj%gridname, trim(nudgeobj%mod_coefficient_name), nudgeobj%model_coefficient, nudge_time, &
+   override=nudge_override(3), is_in=is_in, ie_in=ie_in, js_in=js_in, je_in=je_in)
+
+   nudgeobj%nudge_output = nudgeobj%model_input * nudgeobj%model_coefficient + nudgeobj%observed_data * nudgeobj%observation_coefficient
+   nudge_success = ALL(nudge_override)
+end subroutine data_nudge_3d
+!===============================================================================================
+! <SUBROUTINE NAME="data_nudge_2d">
+!   <DESCRIPTION>
+! This routine performs data nudge for a variable, its observations, and the associated coefficients.
```
Comment (thomas-robinson) [inline data_override/data_override.F90]: Doxygen comments

File: data_override/data_override.F90 (hunk 2 above)
Comment (GFDL-Eric) [reply data_override/data_override.F90]: function name changed (also added nudge2d_construct)

File: data_override/data_override.F90 (hunk 3 above)
Comment (GFDL-Eric) [reply data_override/data_override.F90]: changed as suggested

File: data_override/data_override.F90 (hunk 5 above)
Comment (GFDL-Eric) [reply data_override/data_override.F90]: yerp

File: test_fms/data_override/data_table_base (hunk 6 above)
Comment (GFDL-Eric) [reply test_fms/data_override/data_table_base]: They are required for the 3D / nudge test (can/will separate out into different commit if desired).

File: test_fms/data_override/test_data_override2.sh (hunk 7 above)
Comment (GFDL-Eric) [reply test_fms/data_override/test_data_override2.sh]: fixed, also added skip because the appropriate INPUT files aren't in the CI

**Comment by thomas-robinson** [comment]
@GFDL-Eric did you push the updates?

File: data_override/data_override.F90 (hunk 1 above)
Comment (GFDL-Eric) [reply data_override/data_override.F90]: Changed to Doxygen (comments welcome)

File: data_override/data_override.F90 (hunk 4 above)
Comment (GFDL-Eric) [reply data_override/data_override.F90]: Changed to Doxygen (comments welcome)

File: data_override/data_override.F90 (hunk 8 above)
Comment (GFDL-Eric) [reply data_override/data_override.F90]: Changed to Doxygen (comments welcome)

**Comment by GFDL-Eric** [comment]
Pushed the changes... let me know what you think!

**Comment by bensonr** [comment]
@GFDL-Eric I didn't think we were implementing the specific nudging algorithm within FMS.  I recall the plan was to provide only the 3D data_override capability only within FMS and demonstrate to the users via examples how to use it to perform the requested nudging. 

@jwdGFDL comments?

**Comment by GFDL-Eric** [comment]
I've made a new branch that only has the 3D override implementation as discussed earlier (I'm keeping the nudge branch around to use as the basis for the nudge implementation portion of the presentation). Should I go ahead and close this PR and open a new one with that branch?

**Comment by wrongkindofdoctor** [comment]
@GFDL-Eric @thomas-robinson Yes, please close this PR and open a new one with just the changes for 3D data override that we agreed to support.

**Comment by thomas-robinson** [comment]
@GFDL-Eric does #289 have the changes you want to back out?  If it does, then I need to close that and update.

**Comment by GFDL-Eric** [comment]
@thomas-robinson it does. Sorry! The new (fresh) commit will only have a one line change to time_interp_external2, but the test_fms/data_override directory has several changes. data_override.F90 doesn't actually change!

**Pull request closed without merging**